### DISCUSSION TITLE
LanguageSwitcher does not switch language (Interactive Server / Interactive Webassembly) because of httpOnly culture cookie.

### DISCUSSION THREAD

**Discussion by maurocavallin** [oqtane/oqtane.framework#4703] [General] [answered]
![image](https://github.com/user-attachments/assets/a5183bd5-b26f-4ab3-9e6f-5b2b71029fb0)
 
 ### Problem: Language Switcher Not Working

When I select a different language from the dropdown in the Language Switcher, nothing happens. 
It happens in Interactive/Server and in Interactive/WebAssembly

The code behind is correctly calling `SetCultureAsync`:

```csharp
// Oqtane.Client\Themes\Controls\Theme\LanguageSwitcher.razor
private async Task SetCultureAsync(string culture)
{
    if (culture != CultureInfo.CurrentUICulture.Name)
    {
        var localizationCookieValue = CookieRequestCultureProvider.MakeCookieValue(new RequestCulture(culture));
        var interop = new Interop(JSRuntime);
        await interop.SetCookie(CookieRequestCultureProvider.DefaultCookieName, localizationCookieValue, 360);
        NavigationManager.NavigateTo(NavigationManager.Uri, forceLoad: true);
    }
}
```  
 
also the javasript code is ok:

```csharp
 // Oqtane.Server\wwwroot\js\interop.js
 setCookie: function (name, value, days) {
     var d = new Date();
     d.setTime(d.getTime() + (days * 24 * 60 * 60 * 1000));
     var expires = "expires=" + d.toUTCString();
     document.cookie = name + "=" + value + ";" + expires + ";path=/";
 },
```  
 
The problem I found is that document.cookie is not writable because the language cookie '.AspNetCore.Culture' is an httpOnly cookie: it's not accessible (nor writeable) by javascript.

My workaround for now is to set to false the cookie HttpOnly property:

```csharp
// Oqtane.Server\Components\App.razor
private void SetLocalizationCookie(string culture)
{
    var cookieOptions = new Microsoft.AspNetCore.Http.CookieOptions
    {
        Expires = DateTimeOffset.UtcNow.AddYears(1),
        SameSite = Microsoft.AspNetCore.Http.SameSiteMode.Lax, // Set SameSite attribute
        Secure = true, // Ensure the cookie is only sent over HTTPS
        HttpOnly = false // Optional: Helps mitigate XSS attacks    <----- I CHANGED THIS to false
    };

    Context.Response.Cookies.Append(
        CookieRequestCultureProvider.DefaultCookieName,
        CookieRequestCultureProvider.MakeCookieValue(new RequestCulture(culture)),
        cookieOptions
    );
```

**Comment by sbwalker** [comment]
@thabaum this seems to be a regression error caused by the recent PR related to cookies.

https://github.com/oqtane/oqtane.framework/pull/4493

**Comment by hishamco** [comment]
This option changes the cookie settings to be used with HTTP only. AFAIK everything works fine

**Comment by hishamco** [comment]
> @thabaum this seems to be a regression error caused by the recent PR related to cookies.

@sbwalker could you refer me to the PR, so I can fix the issue

**Reply by thabaum** [reply]
@hishamco https://github.com/oqtane/oqtane.framework/pull/4493

I believe he is referring to this ?

**Reply by thabaum** [reply]
@sbwalker i am reviewing as well to see.  Maybe it does not need set and can just be left out as the main purpose of the original PR was for `SameSite = Microsoft.AspNetCore.Http.SameSiteMode.Lax, // Set SameSite attribute` to fix that issue. 

Does removing this setting fix the issue and just let Oqtane choose whats best.

`HttpOnly = true // Optional: Helps mitigate XSS attacks` this is "optional" so maybe we can safely remove this line and it works ok for @maurocavallin issue discussed here?  I would need to investigate more.

**Reply by maurocavallin** [reply]
@thabaum 
I think the best is to explicitly `HttpOnly = false`, yes the same effect of having a false value would be to omit the line. But having an eplicit line of code (maybe with a comment) will be a memento for devs that we need it to be set to false because this cookie need to be manipulated on the javascript side.

I'm talking only about this specific language cookie, no other cookie.

**Reply by thabaum** [reply]
@maurocavallin thank you.  I will do some testing to see if I can replicate this today along with research more about this issue so I can get a full understanding of it.  Thanks again for your reply I have been a bit busy lately and will share what I find.

**Reply by thabaum** [reply]
I see we have this same setting from the PR https://github.com/oqtane/oqtane.framework/pull/4493 for the visitor cookie.  This issue is only with the localization cookie?

Visitor cookie adds these options:
```
                        SameSite = Microsoft.AspNetCore.Http.SameSiteMode.Lax, // Set SameSite attribute
                        Secure = true, // Ensure the cookie is only sent over HTTPS
                        HttpOnly = true // Optional: Helps mitigate XSS attacks
```

Culture cookie adds these options:

```
                        Expires = DateTimeOffset.UtcNow.AddYears(1),
                        SameSite = Microsoft.AspNetCore.Http.SameSiteMode.Lax, // Set SameSite attribute
                        Secure = true, // Ensure the cookie is only sent over HTTPS
                        HttpOnly = true // Optional: Helps mitigate XSS attacks
```


This is what was mainly changed
https://github.com/oqtane/oqtane.framework/pull/4493/files

And the HttpOnly was to mitigate XSS attacks, which I am just trying to review how this type of attack can be handled in Oqtane if we set to false, or do we even worry about these types of attacks?  I am reviewing if there is an alternative or do we just set this to false as seems to be your recommendation and just do this in areas that wont allow Oqtane to function properly.

Right now I am hearing only the localization cookie is having an issue.  I would like to know more about why it is an issue.

**Reply by thabaum** [reply]
`Cookie “.AspNetCore.Culture” does not have a proper “SameSite” attribute value. Soon, cookies without the “SameSite” attribute or with an invalid value will be treated as “Lax”. This means that the cookie will no longer be sent in third-party contexts. If your application depends on this cookie being available in such contexts, please add the “SameSite=None“ attribute to it. To know more about the “SameSite“ attribute, read https://developer.mozilla.org/docs/Web/HTTP/Headers/Set-Cookie/SameSite`

I see there is also a samesite attribute issue still existing for the culture cookie in 5.3.4.

**Reply by thabaum** [reply]
![image](https://github.com/user-attachments/assets/363c7d0e-f28c-4add-961e-462729f12011)
I also see the cookie being loaded and set, along with the browser remembering what cookie setting was selected.

![image](https://github.com/user-attachments/assets/43741412-29c0-458a-bd02-7b5b1141b268)

**Reply by thabaum** [reply]
Sorry I had set in the other area, i am still reviewing.

**Reply by thabaum** [reply]
seems like a reload may need to happen when changing as you don't see the change until a refresh.

however, your issue is not reproduced as I see the same results when changing the variable.  The language pack itself may not be contributed to.

**Reply by thabaum** [reply]
![image](https://github.com/user-attachments/assets/899a6bff-648f-418d-9e37-573b0b797d7e)

![image](https://github.com/user-attachments/assets/028c697e-d650-467a-b15d-c6dda1bd9d18)

@maurocavallin This is the only place I see the language actually change, and it works the same no matter how I change the httpOnly setting.  This is in Visual Studio. 

How do I reproduce what you are experiencing.  My apologies as I do not work with the language packs in Oqtane.  I still would love to reproduce the issue.

**Reply by thabaum** [reply]
Also if you could please report it as an issue with steps to reproduce that would be a great help!  If not and dont have time would rather continue this discussion here that is fine as well as I will try to work on replicating the issue as soon as I fully understand how to.  If you do decide to create an actual issue we can track for the project you can also reference this discussion for additional information. Thank you.

**Reply by thabaum** [reply]
The issue I believe is it needs to reload to see the change.  Please verify because I think that when changing the setting it stays the same, until you refresh manually.  So this would be a different issue.

`Oqtane.Client\Themes\Controls\Theme\LanguageSwitcher.razor` page reload/refresh on change does not appear to be happening to refresh the UI is the issue I am experiencing.

**Reply by maurocavallin** [reply]
Hi @thabaum sorry for the delay.

Just the last hint (before submitting a issue, and not waste more of your time :-) ):

I'm running from VisualStudio and IIS Express. Latest dev branch.

I experience my issue in two cases:
Render Mode: Interactive Interactivity: Server
Render Mode: Interactive Interactivity: Webassembly

With Render Mode: Static I have no problems.

Before the test, to start from clean, I suggest those operations in sequence:
- change the interactivity mode (i.e.: Render Mode: Interactive Interactivity: Server)
- wipe the ".AspNetCore.Culture" coockie from browser
- restart the application

**Reply by thabaum** [reply]
@maurocavallin thank you for the clarification.  I was able to reproduce this issue with interactive mode and you are correct that setting this httpOnly cookie option to false resolves.

Why does it work in static and not interactive?  Due to SignalR?  I put a PR up recently working in this area and javascript I know I need to go back through and rethink as it was just to show enhancing the SameSite of cookies created via JS interop.  My bad testing via static and not interactive as I know a lot use interactive still.

Since it works in static we should possibly check which render mode Oqtane is running and set this cookie option to true for static and false for interactive render modes if we cannot get it to work properly set to true.

I have tested this on production and same results.  Works OK in static but need to refresh to see the UI update which is another issue.  Interactive mode there is an issue with this cookie.

I will try testing some stuff tomorrow to see if we can get this to work in interactive mode.  Bottom line currently is Static mode works, interactive does not.

It would be most secure to have it set to true, but since it does not work in interactive mode currently we may need to set to false for this mode, but I would still like to keep it set to true for static if we can. It would limit phishing and social engineering attacks from my understanding.

The visitor tracking cookie would also need to be reviewed I would think as well.

**Comment by thabaum** [comment]
@sbwalker https://github.com/oqtane/oqtane.framework/issues/4714 This issue has been resolved setting HttpOnly=false for all modes, however I added a comment relating in the PR for static mode since it did not have an issue maybe HttpOnly would be best to have set to true for this case?

**Reply by sbwalker** [reply]
It would be best to have 1 consistent setting for this cookie, as there is the possibility that a person will switch render modes in their site - and if they started with Static render mode and later changed to Interactive - their users would have issues because of the HttpOnly cookie setting.

**Reply by sbwalker** [reply]
The fact that this cookie is HttpOnly=false is not an issue as this cookie does not contain any sensitive/private information - it simply contains a culture code

**Reply by thabaum** [reply]
I believe the cookie can still be used for social engineering and phishing from my understanding.

**Reply by thabaum** [reply]
#4730 is an example of what I was thinking for a fix. 

But I also want to point out [line 435](https://github.com/oqtane/oqtane.framework/blob/ec6a6d6653e07ef0849dd9791da01a609ac87d1a/Oqtane.Server/Components/App.razor#L435) with the visitor cookie in the app.razor file.  Is this working ok in interactive mode?

**Reply by sbwalker** [reply]
The visitor cookie is never accessed from razor components - it only exists server-side

**Reply by thabaum** [reply]
I have updated #4730 and seems to work OK, however it should be tested further with multiple tenants in different modes.

**Reply by sbwalker** [reply, marked as answer]
If a cookie has HttpOnly disabled it does not mean your site is vulnerable to attack. HttpOnly is only an extra layer of defense (for those browsers that actually support it). XSS requires an initial attack vector - the injection of executable JavaScript into a page. If your site allows a malicious user to inject JavaScript then they could include a script which can read the value of cookies. HttpOnly prevents scripts from reading cookie values. So it's useful as an additional safeguard for authentication cookies or cookies which store other confidential information. But if your site allows malicious scripts to be injected then you obviously have much bigger problems to worry about than just cookies.
